content/data-loading.md

```diff
@@ -0,0 +1,556 @@
+
+
+## Patterns for data loading
+
+Across Meteor applications, there are some common patterns of data loading and and management on the client side that are worth knowing. We'll go into more detail about some of these in the {% post_link ui-ux "UI/UX Article" %}.
+
+### Subscription readiness
+
+A key thing to keep in mind is that a subscription will not instantly provide it's data. There'll be a latency between subscribing to the data on the client and it arriving from the publication on the server (and keep in that this time may be a lot longer for your users in production that for you locally in development!)
+
+Although the Tracker system means you often don't *need* to think too much about this in building your apps, usually if you want to get the user experience right, you'll need to know when the data is ready. 
+
+To find that out, `Meteor.subscribe()` and (`this.subscribe()` in templates) returns a "subscription handle", which contains a reactive data source called `.ready()`:
+
+```js
+const handle = Meteor.subscribe('lists/public');
+Tracker.autorun(() => {
+  console.log(`Handle is ${handle.ready() ? 'ready' : 'not ready'}`);  
+});
+```
+
+We can use this information to be more subtle about when we try and show data to users, and when we show a loading screen.
+
+### Changing subscription arguments
+
+We've seen an example already of using an `autorun` to re-subscribe when the (reactive) arguments to a subscription change. It's worth digging in a little more detail to understand what happens in this scenario.
+
+```js
+Template.listsShowPage.onCreated(function() {
+  this.state = new ReactiveDict();
+  this.autorun(() => {
+    this.state.set('listId', FlowRouter.getParam('_id'));
+    this.subscribe('list/todos', this.state.get('listId'));
+  });
+});
+```
+
+In our example, the `autorun` will re-run whenever `this.state.get('listId')` changes, (ultimately because `FlowRouter.getParam('_id')` changes), although other common reactive data sources are:
+
+1. Template data contexts (which you can access reactively with `Template.currentData()`)
+2. The current user status (`Meteor.user()` and `Meteor.loggingIn()`)
+3. The contents of other application specific client data stores.
+
+Technically, what happens when one of these reactive sources changes is the following:
+
+1. The reactive data source *invalidates* the autorun computation (tells it to re-run itself)
+2. The subscription detects this, and given that anything is possible in next computation run, marks itself for destruction.
+3. The computation re-runs, with `.subscribe()` being re-called either with new or different arguments.
+4a. If the subscription is run with the *same arguments* then the "new" subscription discovers the old "marked for destruction" subscription that sitting around, with the same data already ready, and simply reuses that.
```

Comment:
This should be like an indented list or something, right? AFAIK markdown won't recognize this markup.